lib/Backend/GlobOptArrays.cpp

```diff
@@ -0,0 +1,1896 @@
+                    jitOptimizedBaseOpnd->GetType(),
+                    instr->m_func),
+                instr->m_func);
+    }
+    else
+    {
+        loadObjectArray = nullptr;
+    }
+
+    IR::Instr *const loadHeadSegment =
+        IR::Instr::New(
+            Js::OpCode::LdIndir,
+            headSegmentOpnd,
+            IR::IndirOpnd::New(
+                loadObjectArray ? headSegmentOpnd : jitOptimizedBaseOpnd,
+                Lowerer::GetArrayOffsetOfHeadSegment(baseValueType),
+                headSegmentOpnd->GetType(),
+                instr->m_func),
+            instr->m_func);
+
+    if (hoistHeadSegmentLoadOutOfLoop)
+    {
+        Assert(!(isLikelyJsArray && hoistHeadSegmentLoadOutOfLoop->jsArrayKills.KillsArrayHeadSegments()));
+
+        TRACE_PHASE_INSTR(
+            Js::ArraySegmentHoistPhase,
+            instr,
+            _u("Hoisting array segment load out of loop %u to landing pad block %u\n"),
+            hoistHeadSegmentLoadOutOfLoop->GetLoopNumber(),
+            hoistHeadSegmentLoadOutOfLoop->landingPad->GetBlockNum());
+
+        TESTTRACE_PHASE_INSTR(Js::ArraySegmentHoistPhase, instr, _u("Hoisting array segment load out of loop\n"));
+
+        if (loadObjectArray)
+        {
+            InsertInstrInLandingPad(loadObjectArray, hoistHeadSegmentLoadOutOfLoop);
+        }
+
+        InsertInstrInLandingPad(loadHeadSegment, hoistHeadSegmentLoadOutOfLoop);
+    }
+    else
+    {
+        if (loadObjectArray)
+        {
+            loadObjectArray->SetByteCodeOffset(instr);
+            insertBeforeInstr->InsertBefore(loadObjectArray);
+        }
+
+        loadHeadSegment->SetByteCodeOffset(instr);
+        insertBeforeInstr->InsertBefore(loadHeadSegment);
+        instr->loadedArrayHeadSegment = true;
+    }
+}
+
+void GlobOpt::ArraySrcOpt::Optimize()
+{
+    Assert(instr != nullptr);
+
+    if (!CheckOpCode())
+    {
+        return;
+    }
+
+    Assert(!(baseOwnerInstr && baseOwnerIndir));
+    Assert(!needsHeadSegmentLength || needsHeadSegment);
+
+    if (baseOwnerIndir && !globOpt->IsLoopPrePass())
+    {
+        // Since this happens before type specialization, make sure that any necessary conversions are done, and that the index
+        // is int-specialized if possible such that the const flags are correct.
+        globOpt->ToVarUses(instr, baseOwnerIndir, baseOwnerIndir == instr->GetDst(), nullptr);
+    }
+
+    if (isProfilableStElem && !globOpt->IsLoopPrePass())
+    {
+        // If the dead-store pass decides to add the bailout kind IR::BailOutInvalidatedArrayHeadSegment, and the fast path is
+        // generated, it may bail out before the operation is done, so this would need to be a pre-op bailout.
+        if (instr->HasBailOutInfo())
+        {
+            Assert(instr->GetByteCodeOffset() != Js::Constants::NoByteCodeOffset && instr->GetBailOutInfo()->bailOutOffset <= instr->GetByteCodeOffset());
+
+            const IR::BailOutKind ba
```

Comment:
Seems like this all should be under #if ENABLE_DEBUG_CONFIG_OPTIONS as this is just tracing.